bin/node/runtime/src/lib.rs

```diff
@@ -1325,15 +1332,46 @@ impl_runtime_apis! {
 		}
 	}
 
+	#[cfg(feature = "try-runtime")]
+	impl frame_try_runtime::TryRuntime<Block> for Runtime {
+		fn on_runtime_upgrade(target: frame_try_runtime::Target) -> Weight {
+			frame_support::debug::RuntimeLogger::init();
+
+			let weight = match target {
+				frame_try_runtime::Target::All => {
+					frame_support::debug::info!("Dry-running all on-runtime-upgrades.");
+					Executive::try_runtime_upgrade()
+				},
+				frame_try_runtime::Target::Pallet(name) => {
+					let name = sp_std::str::from_utf8(&name).unwrap();
+					frame_support::debug::info!("Dry-running on-runtime-upgrade of {}.", name);
+
+					frame_try_runtime::match_pallet_on_runtime_upgrade!(name,
+						System, Utility, Babe, Timestamp, Authorship, Indices, Balances,
+						TransactionPayment, Staking, Session, Democracy, Council,
+						TechnicalCommittee, Elections, TechnicalMembership, Grandpa, Treasury,
+						Contracts, Sudo, ImOnline, AuthorityDiscovery, Offences, Historical,
+						RandomnessCollectiveFlip, Identity,  Society, Recovery, Vesting, Scheduler,
+						Proxy, Multisig, Bounties, Tips, Assets, Mmr, Lottery,
+					)
+					// TODO: ^^ this can be done better, or generated by construct runtime at the
```

Comment:
And `AllModules`?

Reply:
See: https://github.com/paritytech/polkadot/pull/2447#discussion_r577397435

Reply:
I agree it is difficult to just use AllModules with PalletInfo,
We could add a method `OnRuntimeUpgrade::try_runtime_upgrade_filtered(name: &str)` and the pallet macro would implemented it. https://github.com/paritytech/polkadot/pull/2447#discussion_r578313731
But all in all I kind of wonder of the intereset here. Why not make the bot simply trigger all runtime upgrade ?
compared to benchmark it shouldn't be too long I think.

Reply:
`try_runtime_upgrade_filtered ` would work but it would be a bad choice in the sense that we would have to alter the clean API of `OnRuntimeUpgrade` for the limitations of `construct_runtime` and other components. 

 > But all in all I kind of wonder of the intereset here. Why not make the bot simply trigger all runtime upgrade ?
compared to benchmark it shouldn't be too long I think.

Indeed, it is not a matter of speed but rather having less noise and false negative. Might be rare but I can predict taht it will be useful. In a PR, we test per-pallet, and in release, we test all of them. 

Overall, if it was up to me, I would let this be for the current PR, and make an issue for it to implement `type AllModulesList` as I explained in the polkadot comment linked above in the future.

Reply:
> Indeed, it is not a matter of speed but rather having less noise and false negative

I don't see a false negative here actually.

> UPDATE: now I see your diff and if we feature gate it... well it is better like that. Maybe we can do that. But what if next we want to exeucte the OnIntialize of a pallet for testing (I actually need this at some point)? This will then require all traits to be updated. I think having PalletCommonTrait and a Vec<Box<dyn PalletCommonTrait>> could work better.

Why do you need to execute an individual OnInitialize ? I really feel like in the runtime pallet are interconnected, having only one pallet initialized may be not working

Reply:
> I think having PalletCommonTrait and a Vec<Box<dyn PalletCommonTrait>> could work better.

I'm a bit skeptical, OnRuntimeUpgrade cannot be made into an object, I don't think we can do `dyn` stuff on them.
For having a trait object, methods needs to take `self` (or ref).

generally speaking every pallets are different type so it is hard to put them into a single Vec.

Reply:
For example, in testing the new multi-phase, I was thinking it would be good to somehow simulate its existent using a new method in the same runtime api that would just call the `OnInitialize` of all or one pallet. 

Then, 
for range of block `b`: 
   - scrape state of polkadot block `b`.
   - concatenate it with the state of pallet-multi-phase (or its genesis state, if no previous state is recorded)
   - call `on_initialize` of the new pallet. 
   - save the new state of somewhere in a hashmap for the next round 
   
This essentially simulates the runtime if this pallet was also inside it, and I guess you are right, in such cases, we should probably execute all `on_initialize`s since there could be dependencies between them.

That being said I will be happier to remove the functionality for executing things single pallet if deemed not a good approach, rather than adding the `on_runtime_upgrade_filtered`. We can surely start with all pallets, and over time add the feature for a single pallet if the use case appears, in hopefully a nicer way.

Reply:
> > I think having PalletCommonTrait and a Vec<Box> could work better.
> 
> I'm a bit skeptical, OnRuntimeUpgrade cannot be made into an object, I don't think we can do `dyn` stuff on them.
> For having a trait object, methods needs to take `self` (or ref).
> 
> generally speaking every `Pallet` are different type so it is hard to put them into a single Vec.

Interesting, I didn't actually tried any of it and assumed it is possible. In this case then let's remove this feature of running a single pallet for now.

Reply:
https://github.com/paritytech/substrate/pull/8038/commits/c8ba546b87b9d8c7ff01633723a74e36d8e5b9e0 Removes this for now.